pkg/terminal/command.go

```diff
@@ -2845,12 +2845,25 @@ func printBreakpointInfo(t *Term, th *api.Thread, tracepointOnNewline bool) {
 	}
 }
 
+var (
+	depth = make(map[int64]int)
+)
+
 func printTracepoint(t *Term, th *api.Thread, bpname string, fn *api.Function, args string, hasReturnValue bool) {
 	if t.conf.TraceShowTimestamp {
 		fmt.Fprintf(t.stdout, "%s ", time.Now().Format(time.RFC3339Nano))
 	}
 
 	if th.Breakpoint.Tracepoint {
+		idx := strings.Index(fn.Name(), "runtime.morestack_noctxt")
```

Comment:
This code isn't going to work correctly if someone sets a tracepoint manually (depth will never be decremented). It will also not work if one of the traced functions doesn't return but panics instead, and the panic is caught by a recover higher on the stack.

Reply:
Hi Alessandro i tried the following tests around panic, in both cases it appears that it works as expected. 

1. Panic does not recover
Code: 
```
package main

import "fmt"

func D(i int) int {
        panic("boom!")
        return i * i * i
}
func C(i int) int {

        return i + 20
}
func B(i int) int {
        d := C(i) + 40
        return d + D(i)
}
func A(i int) int {
        return 10 + B(i)
}
func main() {
        j := 0
        j += A(2)
        fmt.Println(j)
}

```
Trace output: 
```
 [archanaravindar@fedora _fixtures]$ ../dlv trace main.A --follow-calls 4 -e=./leafpanic
> goroutine(1): main.A(2)
 > goroutine(1): main.B(2)
  > goroutine(1): main.C(2)
  >> goroutine(1): => (22)
  > goroutine(1): main.D(2)
   > goroutine(1): runtime.gopanic(interface {}(string) "boom!")
> [unrecovered-panic] runtime.fatalpanic() /home/archanaravindar/Go/Go_master/go/src/runtime/panic.go:1175 (hits goroutine(1):1 total:1) (PC: 0x431024)
Warning: debugging optimized function
	runtime.curg._panic.arg: interface {}(string) "boom!"
  1170:	// fatalpanic implements an unrecoverable panic. It is like fatalthrow, except
  1171:	// that if msgs != nil, fatalpanic also prints panic messages and decrements
  1172:	// runningPanicDefers once main is blocked from exiting.
  1173:	//
  1174:	//go:nosplit
=>1175:	func fatalpanic(msgs *_panic) {
  1176:		pc := getcallerpc()
  1177:		sp := getcallersp()
  1178:		gp := getg()
  1179:		var docrash bool
  1180:		// Switch to the system stack to avoid any stack growth, which

```
2. Panic recovers
Code: 
```
package main

import "fmt"

func D(i int) int {
        defer func() {
                if r:=recover() ; r==nil {
                        fmt.Errorf("function should panic")
                } else {
                        fmt.Printf("successful recovery from panic\n")
                }
        }()
        panic("boom!")
        return i * i * i
}
func C(i int) int {

        return i + 20
}
func B(i int) int {
        d := C(i) + 40
        return d + D(i)
}
func A(i int) int {
        return 10 + B(i)
}
func main() {
        j := 0
        j += A(2)
        fmt.Println(j)
}

```
Trace output: 
```
[archanaravindar@fedora _fixtures]$ ../dlv trace main.A --follow-calls 3 -e=./leafpanic
> goroutine(1): main.A(2)
 > goroutine(1): main.B(2)
  > goroutine(1): main.C(2)
  >> goroutine(1): => (22)
  > goroutine(1): main.D(2)
  > goroutine(1): runtime.morestack_noctxt()
successful recovery from panic
  >> goroutine(1): => (0)
 >> goroutine(1): => (0)
>> goroutine(1): => (62)
>> goroutine(1): => (72)
72
Process 2135496 has exited with status 0

```

Reply:
> This code isn't going to work correctly if someone sets a tracepoint manually (depth will never be decremented).

You are right, as of now, we have implemented this option only for running with dlv at the command line prompt before entering into the dlv shell command environment 
we can still set a tracepoint manually using the trace option from within dlv and it will print a trace for that particular function but we dont support follow-calls option inside dlv yet 
```
[archanaravindar@fedora _fixtures]$ ../dlv exec ./leaf4
Type 'help' for list of commands.
(dlv) trace main.C
Tracepoint 1 set at 0x47b064 for main.C() ./leaf4.go:8
(dlv) continue
> goroutine(1): main.C(2)
>> goroutine(1): => (22)
80
Process 2120528 has exited with status 0
(dlv) q
[archanaravindar@fedora _fixtures]$ ../dlv exec ./leaf4
Type 'help' for list of commands.
(dlv) trace main.B
Tracepoint 1 set at 0x47b0aa for main.B() ./leaf4.go:12
(dlv) continue
> goroutine(1): main.B(2)
>> goroutine(1): => (70)
80
Process 2120556 has exited with status 0
```

Reply:
The problem is that a tracepoint set the the `trace` command doesn't necessarily create the corresponding TraceReturn tracepoints so depth isn't guaranteed to be decremented. Same goes for recovered panics.

Reply:
The case of the panic + recover is interesting however. @archanaravindar as you can see from the example you provided in that case, after the panic is recovered the depth in the output is still the same. We could try and detect a recover, but then how do we know how much to decrement depth by?